File: Data Structures & Algorithms/sliding-window-maximum/submission-2.py
class Solution:
    import collections

    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:

        l = r = 0
        deq = collections.deque()
        res = []

        while r < len(nums):
            while deq and nums[deq[-1]] < nums[r]:
                deq.pop()
            
            deq.append(r)

            if l > deq[0]:
                deq.popleft()
            
            if (r + 1) >= k:
                res.append(nums[deq[0]])
                l += 1
            r += 1
        return res

File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.rs
use std::collections::HashMap;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut res = vec![];
        let max_count = nums.len() + 1;

        let mut counter:HashMap<i32,usize> = HashMap::new();

        for num in nums {
            *counter.entry(num).or_insert(0) += 1;
        }

        let mut buckets:Vec<Vec<i32>> = vec![vec![]; max_count];
        for (&num, &count) in counter.iter() {
            buckets[count].push(num);
        }
        for bucket in buckets.iter().rev(){
            for &num in bucket {
                res.push(num);
                if res.len() == k as usize {
                    return res;
                }
            }
        }
        unreachable!("constraint: there always exists a unique solution")
    }
}
